### DISCUSSION TITLE
Welcome to NBomber Discussions!

### DISCUSSION THREAD

**Discussion by AntyaDev** [VIP-Logic/NBomber#274] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by tienle488** [comment]
Hi.  Where can i find documentation on how to use the NBomber.Performance.sln?

Thanks,
Tien

**Reply by AntyaDev** [reply]
Hi @tienle488 ,
This project exists in the repository to test the internals of NBomber. I am not sure that this is something you need to run as end-user. But if you are interested you can just open the solution and press F5+without debugger.

**Reply by tienle488** [reply]
got it.  thanks, @AntyaDev

**Comment by davkoc** [comment]
Hi, 
Would it be possible to offer a Simulation that will execute only once and without time span limitation? Like Simulation.RunOnce()
I can see that it would be beneficial to be able to write a test and execute it once during development phase/and or as integration test and later, in some deployment pipeline, run the same test but as a load test using other Simulation.

**Reply by AntyaDev** [reply]
Hi @davkoc ,
You can reach it even now.
You need to define CustomSettings for your scenario.
For example:
CustomeSettings: { RunOnce: false}

and within the step, you can check and if the value is true you can stop test execution via context.StopTest()
also, you can understand that step was executed via context.StepInvocation: int

**Reply by davkoc** [reply]
Hi @AntyaDev, 
Thank you for the input - I will try this out :)
However, I still believe that a predefined behavior in terms of a Simulation would be beneficial since the given solution will require all steps to include the check. Plus it will be more clear when reading / configuring the scenario.
Would adding the Simulation.RunOnce() be something you could consider?

**Reply by AntyaDev** [reply]
Can you explain what it gives you?
I mean RunOnce()
What stats are you analyzing? Min, Max, Percentiles

**Reply by davkoc** [reply]
So in our case we want to have tests that developers can run once to validate that a web service works as designed (not to load test it, just for feature validation. The stats is less important (but still interesting) at this stage and this is why we actually want the RunOnce() simulation).
At a later stage in the development process where we ensure that the tested web service works under load/stress we want to use the same test with different simulations. This stage would be a part of a release pipeline for adding new features into the product where non functional aspects of the new feature is tested and ensured.

**Reply by AntyaDev** [reply]
Hi @davkoc, you actually want to have integration tests that you can run as a load test.
It's easy with NBomber.
I even want to write an article about this option.
Basically, you should structure your integration tests in a way that you will separate your user logic from test asserting.

For example, you will have a UserOperation module that contains only functionality to invoke operations like Login, Logout, etc. Then in your Integration tests, you will use this logic and apply asserts based on the result.

The main thing here is that your user operation.Login operation should return HttpResponseMessage.
If you follow this style, you can convert your HttpResponseMessage into NBomber Response.Ok or Fail.
For this, we have an extension method in C# API [NBomber.Http](https://github.com/PragmaticFlow/NBomber.Http/blob/dev/src/NBomber.Http/Api/CSharp.fs#L44)
And here is for [F# API](https://github.com/PragmaticFlow/NBomber.Http/blob/dev/src/NBomber.Http/Api/FSharp.fs#L24)

Here is an [example](https://github.com/PragmaticFlow/NBomber.Http/blob/dev/examples/CSharp/SimpleExample.cs#L23)

Does it make sense?

**Reply by davkoc** [reply]
Hi @AntyaDev I see your point and it makes sense but I just want to re-use the scenario with different load. I am missing the load of a single user executing each step once.

Is this possibly related to "total count load simulation" that you mentioned in Gitter on Sep 10 13:24? And that this will solve my issue/idea?

**Reply by stefan-ce** [reply]
Hi! The suggestion from @davkoc seems very interesting indeed, we are also looking for a solution to code integrations tests in C# and then re-use them as load tests (without rewriting them). Will this be feasible with nBomber in the future @AntyaDev?

**Reply by AntyaDev** [reply]
Hi @stefan-ce @davkoc 
Can you please check this blog: https://nbomber.com/blog/convert-integration-tests-to-load-tests
Does it make sense?

**Reply by davkoc** [reply]
Hi @AntyaDev 
I have read the blog but it does not really solve the issue or problem I am facing. Also looked into your previously suggested solution with  context.StopTest() and context.StepInvocation. That too does not really solve the problem because there is time-related issues doing it this way.

To re-phrase the problem picture a bit...
NBomber performs its test under a specific time using various loads. I want to skip the time aspect and say "run the tests once for one user" - regardless of the time aspect. 
We want to use it to as  way to exercise all endpoints in the system to ensure nothing broke. Not to collect metrics other than to see everything went ok. We will then use the same tests to later on do load- an performance tests and at that point metrics will be interesting as well as the time aspect.

**Reply by AntyaDev** [reply]
@davkoc I got your point.
> We want to use it to as way to exercise all endpoints in the system to ensure nothing broke.

It sounds like you don't need a load test for this but rather PING functionality.
What if tomorrow you need to check some contract schema or make assert based on response for specific endpoint?

My point is that load tests and integration tests have different goals. In my blog post, I described how you can reuse some logic of your integration tests for load tests and it's different from what you propose. My technique allows you to evolve your integration tests separately from load tests including assertions logic. It's two different animals.

**Reply by stefan-ce** [reply]
Hi @AntyaDev. I can in general agree with you about the differences, but there are valid scenarios when they are the same as @davkoc  pointed out. We have been using Visual Studio Load test with the same setup using the same C# tests as integration tests, load tests and stress tests successfully for a couple of years.

The only difference is that integration tests are run once, the load test are running with a wait time and reasonable number of VU:s and finally the stresstests with no wait time and more VUs. But as MS have sunsetted Visual Studio Load test we are looking into other options to keep writing load tests in C#, and nBomber looks promising - if it only had run once. :-) 

**Background**: our use case is a high volume (thousands) of API:s for a large system, and we need to load test each and every API to find the bottlenecks, not the usual load test use case with just a chosen number of the most important scenarios. To be able to handle this volume we need to reuse the integration tests written by the developers of the API (that uses C#/.Net).

I fully understand that a product can't support each and every use case, but I was happy to read your blog about integration tests and hoped for support for our scenario now or in the future.

**Comment by Rjfr10** [comment]
Hi, Am looking to make use of the NBomber Grafana dashboard json file supplied. Do you have an example of how to "import" this file using docker?

**Reply by AntyaDev** [reply]
https://github.com/PragmaticFlow/NBomber.Enterprise.Examples

**Comment by prsantin** [comment]
Hello, is there any intention to make it compatible with net6.0-android?
I was able to run into net6.0-android with few changes!
